TST: a failing `table` test (`test_info_preserved_pickle_copy_init`) on CPython 3.14

This test checks conservation of column attributes through various copy-like operators, at least one of which is currently failing.

reprod (requires Python 3.14.0b1 + numpy 1.23dev)
```shell
pytest astropy/table/tests -k test_info_preserved_pickle_copy_init
```

(trimmed) output
```
...

 >                   assert getattr(m2.info, attr) == original
E                   AssertionError: assert dtype('>i4') == dtype('int32')
E                    +  where dtype('>i4') = getattr(name = m\ndtype = int32\nformat = {0}\ndescription = d\nclass = ArrayWrapper\nn_bad = 0\nlength = 4\n, 'dtype')
E                    +    where name = m\ndtype = int32\nformat = {0}\ndescription = d\nclass = ArrayWrapper\nn_bad = 0\nlength = 4\n = <ArrayWrapper name='m' data=[0 1 2 3]>.info
```

I'm not sure exactly which operation is affected yet as I'm reporting all failures at once, but I'll dig into this one in time.
From afar, it seems likely that either nothing is actually broken (dtypes under comparison are actually equivalent), or that the bug is upstream in numpy. Either way, more digging needed.

--- a/astropy/table/table_helpers.py
+++ b/astropy/table/table_helpers.py
@@ -132,7 +132,12 @@ class ArrayWrapper:
     info = ArrayWrapperInfo()
 
     def __init__(self, data, copy=True):
-        self.data = np.array(data, copy=copy)
+        if isinstance(data, ArrayWrapper):
+            # this is done to preserve byteorder through copies
+            arr = data.data
+        else:
+            arr = data
+        self.data = np.array(arr, copy=copy)
         if "info" in getattr(data, "__dict__", ()):
             self.info = data.info
 

--- a/astropy/table/tests/test_mixin.py
+++ b/astropy/table/tests/test_mixin.py
@@ -503,7 +503,9 @@ def test_info_preserved_pickle_copy_init(mixin_cols):
     """
 
     def pickle_roundtrip(c):
-        return pickle.loads(pickle.dumps(c))
+        # protocol=5 matches the default for Python 3.14 and later
+        # and is needed to preserve byteorder (available since Python 3.8)
+        return pickle.loads(pickle.dumps(c, protocol=5))
 
     def init_from_class(c):
         return c.__class__(c)
@@ -518,18 +520,7 @@ def test_info_preserved_pickle_copy_init(mixin_cols):
         for func in (copy.copy, copy.deepcopy, pickle_roundtrip, init_from_class):
             m2 = func(m)
             for attr in attrs:
-                # non-native byteorder not preserved by last 2 func, _except_ for structured dtype
-                if (
-                    attr != "dtype"
-                    or getattr(m.info.dtype, "isnative", True)
-                    or m.info.dtype.name.startswith("void")
-                    or func in (copy.copy, copy.deepcopy)
-                ):
-                    original = getattr(m.info, attr)
-                else:
-                    # func does not preserve byteorder, check against (native) type.
-                    original = m.info.dtype.newbyteorder("=")
-                assert getattr(m2.info, attr) == original
+                assert getattr(m2.info, attr) == getattr(m.info, attr)
 
 
 def check_share_memory(col1, col2, copy):

--- a/astropy/tests/helper.py
+++ b/astropy/tests/helper.py
@@ -126,10 +126,17 @@ def assert_follows_unicode_guidelines(x, roundtrip=None):
             assert eval(repr_x, roundtrip) == x
 
 
-@pytest.fixture(params=[0, 1, -1])
+@pytest.fixture(params=[0, 1, 4, -1])
 def pickle_protocol(request):
     """
-    Fixture to run all the tests for protocols 0 and 1, and -1 (most advanced).
+    Fixture to run all the tests for protocols 0, 1, 4 and -1 (most advanced).
+
+    * 0 is the original "human-readable" protocol
+    * 1 is the oldest binary format
+    * 4 introduced in Python 3.4, has been the default in 3.8-3.13
+    * -1 is the most advanced, which is 5 since Python 3.8, default from 3.14,
+      and the first binary format to preserve byteorder.
+
     (Originally from astropy.table.tests.test_pickle).
     """
     return request.param

--- a/docs/changes/table/18139.bugfix.rst
+++ b/docs/changes/table/18139.bugfix.rst
@@ -1,0 +1,2 @@
+Fixed a bug in ``table.table_helpers.ArrayWrapper`` where byteorder of the
+underlying data was not necessarily preserved through roundtrips.

--- a/docs/cosmology/io/details.rst
+++ b/docs/cosmology/io/details.rst
@@ -19,7 +19,8 @@ object is using the :mod:`pickle` module. This is good for e.g. passing a
    >>> import pickle
    >>> from astropy.cosmology import Planck18
    >>> with open("planck18.pkl", mode="wb") as file:
-   ...     pickle.dump(Planck18, file)
+   ...     # use protocol 5 to ensure byteorder is preserved (not switched to native)
+   ...     pickle.dump(Planck18, file, protocol=5)
    >>> # and to read back
    >>> with open("planck18.pkl", mode="rb") as file:
    ...     cosmo = pickle.load(file)

